pkg/util/grpc/stats/grpc_stats_test.go

```diff
@@ -0,0 +1,255 @@
+        	            	cortex_grpc_response_size_bytes_bucket{method="/grpc.health.v1.Health/Check",le="5.24288e+07"} 1
+        	            	cortex_grpc_response_size_bytes_bucket{method="/grpc.health.v1.Health/Check",le="1.048576e+08"} 1
+        	            	cortex_grpc_response_size_bytes_bucket{method="/grpc.health.v1.Health/Check",le="2.62144e+08"} 1
+        	            	cortex_grpc_response_size_bytes_bucket{method="/grpc.health.v1.Health/Check",le="+Inf"} 1
+        	            	cortex_grpc_response_size_bytes_sum{method="/grpc.health.v1.Health/Check"} 7
+        	            	cortex_grpc_response_size_bytes_count{method="/grpc.health.v1.Health/Check"} 1
+	`), "cortex_grpc_connected_clients", "cortex_grpc_request_size_bytes", "cortex_grpc_response_size_bytes", "cortex_grpc_method_errors_total")
+	require.NoError(t, err)
+
+	closed = true
+	require.NoError(t, conn.Close())
+
+	// Give server little time to update connected clients metric.
+	time.Sleep(1 * time.Second)
+	err = testutil.GatherAndCompare(reg, bytes.NewBufferString(`
+        	            	# HELP cortex_grpc_connected_clients Number of clients connected to gRPC server
+        	            	# TYPE cortex_grpc_connected_clients gauge
+        	            	cortex_grpc_connected_clients 0
+	`), "cortex_grpc_connected_clients")
+	require.NoError(t, err)
+}
+
+func TestGrpcStatsStreaming(t *testing.T) {
+	reg := prometheus.NewRegistry()
+	stats := NewStatsHandler(reg)
+
+	serv := grpc.NewServer(grpc.StatsHandler(stats), grpc.MaxSendMsgSize(10e6), grpc.MaxRecvMsgSize(10e6))
+	defer serv.GracefulStop()
+
+	listener, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	go func() {
+		require.NoError(t, serv.Serve(listener))
+	}()
+
+	frontend.RegisterFrontendServer(serv, &frontendServer{log: t.Log})
+
+	conn, err := grpc.Dial(listener.Addr().String(), grpc.WithInsecure(), grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(10e6), grpc.MaxCallSendMsgSize(10e6)))
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, conn.Close())
+	}()
+
+	fc := frontend.NewFrontendClient(conn)
+
+	s, err := fc.Process(context.Background())
+	require.NoError(t, err)
+
+	for ix := 0; ix < 5; ix++ {
+		req, err := s.Recv()
+		require.NoError(t, err)
+
+		msg := &frontend.ProcessResponse{HttpResponse: &httpgrpc.HTTPResponse{
+			Code:    200,
+			Headers: req.HttpRequest.Headers,
+		}}
+		t.Log("Client sending:", msg.Size())
+		err = s.Send(msg)
+		require.NoError(t, err)
+
+		err = testutil.GatherAndCompare(reg, bytes.NewBufferString(`
+			# HELP cortex_grpc_inflight_requests Number of inflight RPC calls
+			# TYPE cortex_grpc_inflight_requests gauge
+			cortex_grpc_inflight_requests{method="/frontend.Frontend/Process"} 1
+		`), "cortex_grpc_inflight_requests")
+		require.NoError(t, err)
+	}
+	require.NoError(t, s.CloseSend())
+
+	// Wait a second to make sure server notices close.
+	time.Sleep(1 * time.Second)
```

Comment:
Instead of the sleep, what do you think to use the test polling utility until it succeed?